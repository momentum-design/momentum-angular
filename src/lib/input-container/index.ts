export * from './input-container.component';
export * from './input-container.module';
export * from './input-container.service';
