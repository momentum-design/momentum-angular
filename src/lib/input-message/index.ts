export * from './input-message.component';
export * from './input-message.module';
