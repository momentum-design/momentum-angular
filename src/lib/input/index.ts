export * from './input.directive';
export * from './input.module';
export * from './input.service';

