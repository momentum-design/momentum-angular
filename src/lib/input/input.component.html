

<div class="cui-input-group" [ngClass]="wrapperClasses">
  <cui-label *ngIf="label" [label]="label" className="cui-label"></cui-label>

  <input
    *ngIf="!secondaryLabel"
    class="cui-input"
    [ngClass]="inputClasses"
    [(ngModel)]="value"
    [placeholder]="placeholder"
    [readonly]="readOnly"
    [attr.disabled]="disabled ? '' :null"
    [type]="type"
    (blur)="onBlur($event)"
    (focus)="onFocus($event)"
    (keydown)="onKeyDown($event)"
    (mousedown)="onMouseDown($event)"
  />

  <!-- Secondary Label -->
  <div class="cui-label__secondary-label-container" *ngIf="secondaryLabel">
    <input
      class="cui-input"
      [ngClass]="inputClasses"
      [(ngModel)]="value"
      [type]="type"
      [placeholder]="placeholder"
      [readonly]="readOnly"
      [attr.disabled]="disabled ? '' :null"
      (blur)="onBlur($event)"
      (focus)="onFocus($event)"
      (keydown)="onKeyDown($event)"
      (mousedown)="onMouseDown($event)"
      [name]="name"
    />

    <label class="cui-label__secondary-label">
      <span>{{secondaryLabel}}</span>
    </label>
  </div>

  <!-- Error Input -->
  <ng-container *ngIf="errors">
    <cui-input-error
      *ngFor="let error of errors"
      [error]="error"
    >
    </cui-input-error>
  </ng-container>

  <!-- Angular Control Error Validation Message  -->
  <div class="cui-input__messages" role="alert" *ngIf="control.invalid">
    <div class="message">{{errorValid}}</div>
  </div>

  <!-- Helper Text -->
  <cui-input-helper
    *ngIf="inputHelpText"
    [message]="inputHelpText"
  >
  </cui-input-helper>
</div>