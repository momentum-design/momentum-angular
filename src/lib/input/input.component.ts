import { Component, OnInit, Input, OnChanges, SimpleChanges, forwardRef, Output, EventEmitter, Optional, Self} from '@angular/core';
import { NG_VALUE_ACCESSOR, ControlValueAccessor, NgControl } from '@angular/forms';

const cb = () => {};

@Component({
  selector: 'cui-input',
  template: `
    <div class="cui-input-group" [ngClass]="wrapperClasses">
      <cui-label *ngIf="label" [label]="label" className="cui-label" htmlFor="htmlId"></cui-label>

      <ng-container *ngIf="clear; else normal">
        <div class="cui-input__icon-container">
          <input
            *ngIf="!secondaryLabel"
            class="cui-input"
            [ngClass]="inputClasses"
            [(ngModel)]="value"
            [placeholder]="placeholder"
            [readonly]="readOnly"
            [attr.disabled]="disabled ? '' :null"
            [type]="type"
            (blur)="onBlur($event)"
            (focus)="onFocus($event)"
            (keydown)="onKeyDown($event)"
            (mousedown)="onMouseDown($event)"
            [name]="name"
            [attr.id]="htmlId"
          />

          <button
            *ngIf="clear && value !== ''"
            class="cui-button cui-button--36 cui-button--icon"
            type="button"
            aria-label="clear input"
            (click)="handleClear()"
          >
            <span class="cui-button__children" style="opacity: 1;">
              <i class="cui-icon icon icon-clear-active_16 cui-input__icon-clear"></i>
            </span>
          </button>
        </div>
      </ng-container>

      <ng-template #normal>
        <input
          class="cui-input"
          [ngClass]="inputClasses"
          [(ngModel)]="value"
          [type]="type"
          [placeholder]="placeholder"
          [readonly]="readOnly"
          [attr.disabled]="disabled ? '' :null"
          (blur)="onBlur($event)"
          (focus)="onFocus($event)"
          (keydown)="onKeyDown($event)"
          (mousedown)="onMouseDown($event)"
          [name]="name"
          [attr.id]="htmlId"
        />
      </ng-template>

      <!-- Secondary Label -->
      <div class="cui-label__secondary-label-container" *ngIf="secondaryLabel">
        <input
          class="cui-input"
          [ngClass]="inputClasses"
          [(ngModel)]="value"
          [type]="type"
          [placeholder]="placeholder"
          [readonly]="readOnly"
          [attr.disabled]="disabled ? '' :null"
          (blur)="onBlur($event)"
          (focus)="onFocus($event)"
          (keydown)="onKeyDown($event)"
          (mousedown)="onMouseDown($event)"
          [name]="name"
          [attr.id]="htmlId"
        />

        <label class="cui-label__secondary-label" [attr.for]="htmlId">
          <span>{{secondaryLabel}}</span>
        </label>
      </div>

      <!-- Error Input -->
      <ng-container *ngIf="errors">
        <cui-input-error
          *ngFor="let error of errors"
          [error]="error"
        >
        </cui-input-error>
      </ng-container>

      <!-- Angular Control Error Validation Message  -->
      <div class="cui-input__messages" role="alert" *ngIf="control.invalid">
        <div class="message">{{errorValid}}</div>
      </div>

      <!-- Helper Text -->
      <cui-input-helper
        *ngIf="inputHelpText"
        [message]="inputHelpText"
      >
      </cui-input-helper>
    </div>
  `,
})

export class InputComponent implements ControlValueAccessor {

  private innerValue: any = '';

  private onTouchedCallback: () => void = cb;
  private onChangeCallback: (_: any) => void = cb;

  get value(): any {
    return this.innerValue;
  };

  set value(v: any) {
    if (v !== this.innerValue) {
      this.innerValue = v;
      this.onChangeCallback(v);
    }
  }

  onBlur(event) {
    this.handleBlur.emit()
    event.stopPropagation();
  }

  onFocus(event) {
    if(this.disabled) {
      event.stopPropagation();
      return;
    }
    this.handleFocus.emit();
  }

  onKeyDown(event){
    this.handleKeyDown.emit();
  }

  onMouseDown(event){
    if(this.disabled) {
      event.stopPropagation();
      return;
    }
    this.handleMouseDown.emit();
  }

  writeValue(value: any) {
    if (value !== this.innerValue) {
      this.innerValue = value;
    }
  }

  registerOnChange(fn: any) {
    this.onChangeCallback = fn;
  }

  registerOnTouched(fn: any) {
    this.onTouchedCallback = fn;
  }

  private determineErrorType = array => {
    return array.reduce((agg, e) => {
      return agg === 'error' ? agg : e.type || '';
    }, '');
  };

  private filterErrorsByType = (array, value) => {
    return array.reduce(
      (agg, e) => (e.type === value ? agg.concat(e.error) : agg),
      []
    );
  };

  get hasRequiredError() {
    return this.control.hasError('required');
  }

  get errorValid() {
    // if (this.control.pristine) {
    //   return '';
    // }
    for(let error in this.errorObj) {
      if (this.control.errors[error]) {
        return this.errorObj[error]
      }
    }
    return '';
  }

  /** @prop Optional css class string | ''  */
  @Input() public class: string = '';
  /** @prop optional button to clear input text */
  @Input() public clear: boolean = false;
  /** @prop Sets the disabled attribute of the Input | false */
  @Input() public disabled: boolean = false;
  /** @prop Unique HTML ID used for tying label to HTML input for automated testing */
  @Input() public htmlId: string = '';
  /** @prop Array of objects with error type and error message */
  @Input() public errorArr: any[];
  /** @prop Input css class name string */
  @Input() public inputClass: string = '';
  /** @prop Help Text to appear under the input | '' */
  @Input() public inputHelpText: string = '';
  /** @prop Overall input group size | '' */
  @Input() public inputSize: string = '';
  /** @prop Input label text | '' */
  @Input() public label: string = '';
  /** @prop Placeholder text to display when Input is empty | '' */
  @Input() public placeholder: string = "";
  /** @prop Determines if Input can be edited | false */
  @Input() public readOnly: boolean = false;
  /** @prop Secondary Input label | ''  */
  @Input() public secondaryLabel: string = '';
  /** @prop Input color theme | '' */
  @Input() public theme: string;
  /** @prop Input type | 'text' */
  @Input() public type: string = "text";
  /** @prop Sets the attribute name to the input element | '' */
  @Input() public name: string = '';
  /** @prop Optional error messages object with angular validators | {} */
  @Input() public errorObj: object = {};

  /** @prop function when clicked outside of input */
  @Output() handleBlur: EventEmitter<any> = new EventEmitter();
  /** @prop function when input is focused */
  @Output() handleFocus: EventEmitter<any> = new EventEmitter();
  /** @prop function when key down on input */
  @Output() handleKeyDown: EventEmitter<any> = new EventEmitter();
  /** @prop function when mouse down on input */
  @Output() handleMouseDown: EventEmitter<any> = new EventEmitter();

  public errorType;
  public errors;

  constructor(@Optional() @Self() public control: NgControl) {
    if (this.control) {
      this.control.valueAccessor = this;
    }
  }

  ngOnInit() {}

  ngOnChanges(changes: SimpleChanges){
    if(changes.errorArr){
      let changedArr = changes.errorArr.currentValue

      if(changedArr.length > 0){
        this.errorType = this.determineErrorType(changedArr);
        this.errors = this.errorType && this.filterErrorsByType(changedArr, this.errorType);
      }
    }
  }

  get wrapperClasses() {
    return {
      [this.inputSize + ' columns']: this.inputSize,
      'read-only': this.readOnly,
      'disabled': this.disabled,
      'error': Object.keys(this.errorObj).length > 0 && this.control.invalid && this.control.dirty,
      ['cui-input-group--' + this.theme]: this.theme,
      [this.errorType]: this.errorType,
      [this.class]: this.class
    };
  }

  get inputClasses() {
    return {
      [this.inputClass]: this.inputClass,
      'read-only': this.readOnly,
      'disabled': this.disabled,
      'dirty': this.value
    };
  }

  handleClear(){
    this.value = ""
  }
}

/**
* @component input
* @section default
* @angular
*

<div>
  <cui-input
    [(ngModel)]="dataModel"
    inputSize="small-5"
    label="Default Input"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section error
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    label="Error (Error) Input"
    [errorArr]="[{error: 'This is where the error message would be.', type: 'error'}]"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section warning
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    label="Error (Warning) Input"
    [errorArr]="[{error: 'This is where the success message would be.', type: 'warning'}]"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section success
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    label="Error (Success) Input"
    [errorArr]="[{error: 'This is where the success message would be.', type: 'success'}]"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section disabled
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    [disabled]="true"
    label="Disabled Input"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section read-only
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    [readOnly]="true"
    label="Read Only Input"
    [(ngModel)]="dataModel"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section help-text
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    inputHelpText="Help Text"
    label="Help Text Input"
  >
  </cui-input>
</div>
*/

/**
* @component input
* @section secondary-label
* @angular
*
<div>
  <cui-input
    inputSize="small-5"
    secondaryLabel="Secondary Label"
  >
  </cui-input>
</div>
*/


