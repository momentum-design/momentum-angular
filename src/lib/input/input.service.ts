import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';

@Injectable()
export class InputService {

  private addBeforePadding = new Subject();
  addBeforePadding$ = this.addBeforePadding.asObservable();

  private addAfterPadding = new Subject();
  addAfterPadding$ = this.addAfterPadding.asObservable();

  onPaddingChange(position: string, value: boolean) {
    this[`add${position}Padding`].next(value);
  }
}

