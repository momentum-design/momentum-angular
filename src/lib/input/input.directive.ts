/** @component input */

import {
  DoCheck,
  Directive,
  ElementRef,
  HostBinding,
  Input,
  OnChanges,
  OnDestroy,
  OnInit,
  Optional,
  Self,
  SimpleChanges,
} from '@angular/core';
import { AutofillMonitor } from '@angular/cdk/text-field';
import { takeUntil } from 'rxjs/operators';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { getSupportedInputTypes, Platform } from '@angular/cdk/platform';
import { InputGroupService } from '../input-group';
import { InputService } from './input.service';
import { Subject } from 'rxjs';
import {
  ControlValueAccessor,
  FormControl,
  FormGroupDirective,
  NgControl,
  NgForm,
} from '@angular/forms';

let nextUniqueId = 0;

// Invalid input type. Using one of these will throw an MDInputUnsupportedTypeError.
const MD_INPUT_INVALID_TYPES = [
  'button',
  'checkbox',
  'file',
  'hidden',
  'image',
  'radio',
  'range',
  'reset',
  'submit'
];

function getMDInputUnsupportedTypeError(type: string): Error {
  return Error(`Input type "${type}" isn't supported by mdInput.`);
}

@Directive({
  selector: `input[mdInput], textarea[mdInput]`,
  exportAs: 'mdInput',
  host: {
    '[attr.id]': 'id',
    '[attr.placeholder]': 'placeholder',
    '[attr.aria-describedby]': '_ariaDescribedby || null',
    '[disabled]': 'disabled',
    '[required]': 'required',
    '[readonly]': 'readonly || null',
    '(blur)': 'handleFocusChange(false)',
    '(focus)': 'handleFocusChange(true)',
    '(input)': 'handleInput()'
  },
  providers: [
    {provide: InputGroupService, useExisting: InputDirective}
  ]
})

export class InputDirective implements ControlValueAccessor, DoCheck, InputGroupService<any>, OnChanges, OnDestroy, OnInit {
  private _destroyed = new Subject<void>();
  private _inputValueAccessor: any;
  private _ngControl: any;
  private _parentForm: any;
  private _parentFormGroup: any;
  private _readonly = false;
  protected _disabled = false;
  protected _id: string;
  protected _previousNativeValue: any;
  protected _required = false;
  protected _type = 'text';
  protected _uid = `md-input-${nextUniqueId++}`;
  readonly stateChanges: Subject<void> = new Subject<void>();
  autofilled = false;
  _ariaDescribedby: string;
  _inputAfter = false;
  _inputBefore = false;
  focused: boolean = false;
  errorState: boolean = false;
  status: string;

  /*** @prop Unique HTML ID used for tying label to HTML input | 'md-in-<uniqueId>' */
  @Input()
  get id(): string { return this._id; }
  set id(value: string) { this._id = value || this._uid; }
  @Input()
  /*** @prop Sets the disabled attribute of the Input | false */
  get disabled(): boolean {
    if (this.ngControl && this.ngControl.disabled !== null) {
      return this.ngControl.disabled;
    }
    return this._disabled;
  }
  set disabled(value: boolean) {
    this._disabled = coerceBooleanProperty(value);

    // Browsers may not fire the blur event if the input is disabled too quickly.
    // Reset from here to ensure that the element doesn't become stuck.
    if (this.focused) {
      this.focused = false;
      this.stateChanges.next();
    }
  }
  /*** @prop Placeholder text to display when Input is empty | '' */
  @Input() public placeholder: string = '';
  /*** @prop Determines if Input can be edited | false */
  @Input()
  get readonly(): boolean { return this._readonly; }
  set readonly(value: boolean) { this._readonly = coerceBooleanProperty(value); }
  /*** @prop Determines if Input is required | false */
  @Input()
  get required(): boolean { return this._required; }
  set required(value: boolean) { this._required = coerceBooleanProperty(value); }
  /** @prop Input shape property | '' */
  @Input() public shape: string = '';
  @Input()
  get type(): string { return this._type; }
  set type(value: string) {
    this._type = value || 'text';
    this._validateType();

    // When using Angular inputs, developers are no longer able to set the properties on the native
    // input element. To ensure that bindings for `type` work, we need to sync the setter
    // with the native property. Textarea elements don't support the type property or attribute.
    if (!this._isTextarea() && getSupportedInputTypes().has(this._type)) {
      (this._elementRef.nativeElement as HTMLInputElement).type = this._type;
    }
  }
  /*** @prop Input value | '' */
  @Input()
  get value(): string {
    return this._inputValueAccessor.value;
   }
  set value(value: string) {
    if (value !== this.value) {
      this._inputValueAccessor.value = value;
      this.stateChanges.next();
    }
  }

  protected _neverEmptyInputTypes = [
    'date',
    'datetime',
    'datetime-local',
    'month',
    'time',
    'week'
  ].filter(t => getSupportedInputTypes().has(t));

  @HostBinding('class') get _class(): string {
    return (
      'md-input' +
      `${this._isTextarea() ? ' md-input--multiline' : ''}` +
      `${this.shape ? ` md-input--${this.shape}` : ''}` +
      `${this._inputBefore ? ` md-input--before` : ''}` +
      `${this._inputAfter ? ` md-input--after` : ''}` +
      `${this.focused ? ` md-active` : ''}` +
      `${this.readonly ? ' md-read-only' : ''}` +
      `${this.disabled ? ' md-disabled' : ''}` +
      `${(this._ngControl && this._ngControl.dirty)  ? ` md-dirty` : ''}`
    );
  }

  constructor(
    protected _elementRef: ElementRef<HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement>,
    protected _platform: Platform,
    private _autofillMonitor: AutofillMonitor,
    private inputService: InputService,
    @Optional() @Self() public ngControl: NgControl,
    @Optional() _parentForm: NgForm,
    @Optional() _parentFormGroup: FormGroupDirective
  ) {

    const element = this._elementRef.nativeElement;

    this.id = this.id;
    this._inputValueAccessor = (ngControl && ngControl.valueAccessor) || element;
    this._parentForm = _parentForm;
    this._parentFormGroup = _parentFormGroup;
    this._ngControl = ngControl;

    this.inputService.addBeforePadding$
      .pipe(takeUntil(this._destroyed))
      .subscribe(value => this._inputBefore = coerceBooleanProperty(value));

    this.inputService.addAfterPadding$
      .pipe(takeUntil(this._destroyed))
      .subscribe(value => this._inputAfter = coerceBooleanProperty(value));
  }

  ngDoCheck() {
    if (this.ngControl) {
      // We need to re-evaluate this on every change detection cycle, because there are some
      // error triggers that we can't subscribe to (e.g. parent form submissions). This means
      // that whatever logic is in here has to be super lean or we risk destroying the performance.
      this.updateErrorState();
    }
    // We need to dirty-check the native element's value, because there are some cases where
    // we won't be notified when it changes (e.g. the consumer isn't using forms or they're
    // updating the value using `emitEvent: false`).
    this._dirtyCheckNativeValue();
  }

  ngOnChanges(changes: SimpleChanges): void {
    this.stateChanges.next();
  }

  ngOnInit() {
    if (this._platform.isBrowser) {
      this._autofillMonitor.monitor(this._elementRef.nativeElement).subscribe(event => {
        this.autofilled = event.isAutofilled;
        this.stateChanges.next();
      });
    }
  }

  ngOnDestroy() {
    this._destroyed.next();
    this._destroyed.complete();
    this.stateChanges.complete();

    if (this._platform.isBrowser) {
      this._autofillMonitor.stopMonitoring(this._elementRef.nativeElement);
    }
  }

  protected _isNeverEmpty() {
    return this._neverEmptyInputTypes.indexOf(this._type) > -1;
  }

  /** Checks whether the input is invalid based on the native validation. */
  protected _isBadInput() {
    // The `validity` property won't be present on platform-server.
    const validity = (this._elementRef.nativeElement as HTMLInputElement).validity;
    return validity && validity.badInput;
  }

  protected _isTextarea() {
    return this._elementRef.nativeElement.nodeName.toLowerCase() === 'textarea';
  }

  protected _validateType() {
    if (MD_INPUT_INVALID_TYPES.indexOf(this._type) > -1) {
      throw getMDInputUnsupportedTypeError(this._type);
    }
  }

  get empty(): boolean {
    return !this._isNeverEmpty() && !this._elementRef.nativeElement.value && !this._isBadInput() &&
        !this.autofilled;
  }

  /** Callback for the cases where the focused state of the input changes. */
  handleFocusChange(isFocused: boolean) {
    if (isFocused !== this.focused && (!this.readonly || !isFocused)) {
      this.focused = isFocused;
      this.stateChanges.next();
    }
  }

  handleInput() {
    // This is a noop function and is used to let Angular know whenever the value changes.
    // Angular will run a new change detection each time the `input` event has been dispatched.
    // It's necessary that Angular recognizes the value change, because when floatingLabel
    // is set to false and Angular forms aren't used, the placeholder won't recognize the
    // value changes and will not disappear.
    // Listening to the input event wouldn't be necessary when the input is using the
    // FormsModule or ReactiveFormsModule, because Angular forms also listens to input events.
  }

  onChange: any = () => {};
  onTouch: any = () => {};

  registerOnChange(fn: any): void {
    this.onChange = fn;
  }

  registerOnTouched(fn: any): void {
    this.onTouch = fn;
  }

  setDescribedByIds(ids: string[]) {
    this._ariaDescribedby = ids.join(' ');
  }

  setDisabledState?(isDisabled: boolean): void {
    this.disabled = isDisabled;
  }

  updateErrorState() {
    const oldState = this.errorState;
    const parent = this._parentFormGroup || this._parentForm;
    const control = this.ngControl ? this.ngControl.control as FormControl : null;
    const newState = !!(
      control &&
      control.invalid &&
      (control.dirty || (parent && parent.submitted))
    );

    if (newState !== oldState) {
      this.errorState = newState;
      this.stateChanges.next();
    }
    return;
  }

  writeValue(value: any) {
    if (value !== this._inputValueAccessor.value) {
      this._inputValueAccessor.value = value;
    }
  }

  private _dirtyCheckNativeValue() {
    const newValue = this._elementRef.nativeElement.value;

    if (this._previousNativeValue !== newValue) {
      this._previousNativeValue = newValue;
      this.stateChanges.next();
    }
  }
}

