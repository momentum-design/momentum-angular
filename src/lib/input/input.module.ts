import { CommonModule } from '@angular/common';
import { NgModule, ModuleWithProviders } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { InputDirective } from './input.directive';

@NgModule({
  declarations: [InputDirective],
  imports: [
    CommonModule,
    FormsModule,
    ReactiveFormsModule
  ],
  exports: [InputDirective],
})

export class InputModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: InputModule,
      providers: [],
    };
  }
}
