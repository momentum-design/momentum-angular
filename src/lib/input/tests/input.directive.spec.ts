import { Component } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { IconModule } from '../../icon';
import { InputModule, InputService } from '../../input';
import { InputContainerModule } from '../../input-container';
import { InputMessageModule } from '../../input-message';
import { InputSectionModule } from '../../input-section';

@Component({
  template: `
    <md-input-container
      [isFilled]="isFilled"
      [class]="class"
      [clear]="clear"
      [clearAriaLabel]="clearAriaLabel"
      [containerSize]="containerSize"
      [helpText]="helpText"
      [inputSize]="inputSize"
      [label]="label"
      [messageArr]="messageArr"
      [nestedLevel]="nestedLevel"
      [secondaryLabel]="secondaryLabel"
      [status]="status"
    >
      <md-input-section *ngIf="inputBefore" position='before'>
        <md-icon
          name="clear-active_16"
        ></md-icon>
      </md-input-section>
      <md-input-section *ngIf="inputAfter" position='after'>
        <md-icon
          name="clear-active_16"
        ></md-icon>
      </md-input-section>
      <md-input-message *ngIf="status">{{status}}</md-input-message>
      <input
        mdInput
        id='test-id'
        [disabled]="disabled"
        [placeholder]="placeholder"
        [readonly]="readonly"
        [required]="required"
        [shape]="shape"
        [value]="value"
      />
      <textarea mdInput *ngIf="textArea">test</textarea>
    </md-input-container>`,
})

class MDInputComponent {
  class: string;
  clear: boolean;
  clearAriaLabel: string;
  containerSize: string;
  disabled: boolean;
  helpText: string;
  inputAfter: boolean;
  inputBefore: boolean;
  inputSize: string;
  isFilled: boolean;
  label: string;
  messageArr: Array<{message: string, type: string}>;
  nestedLevel: number;
  placeholder: string;
  readonly: boolean;
  required: boolean;
  secondaryLabel: string;
  shape: string;
  status: string;
  textArea: boolean;
  type: string;
  value: string;
}

describe('InputDirective', () => {
  let component: MDInputComponent;
  let fixture: ComponentFixture<MDInputComponent>;
  let groupEl: HTMLElement;
  let inputEl: HTMLInputElement;
  let wrapperEl: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [MDInputComponent],
      imports: [IconModule, InputMessageModule, InputModule, InputContainerModule, InputSectionModule],
      providers: [InputService]
    }).compileComponents();

    fixture = TestBed.createComponent(MDInputComponent);
    component = fixture.componentInstance;
    groupEl = fixture.debugElement.query(By.css('md-input-container')).nativeElement;
    inputEl = fixture.debugElement.query(By.css('input')).nativeElement;
    wrapperEl = fixture.debugElement.query(By.css('.md-input__wrapper')).nativeElement;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(fixture).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should not overwrite existing id', () => {
    expect(inputEl.id).toBe('test-id');
  });

  it('should render class', () => {
    expect(groupEl.classList.contains('test-input')).toBeFalsy();

    fixture.componentInstance.class = 'test-input';
    fixture.detectChanges();

    expect(groupEl.classList.contains('test-input')).toBe(true);
  });

  it('should render multiline class', () => {
    expect(fixture.debugElement.query(By.css('textarea'))).toBeFalsy();

    fixture.componentInstance.textArea = true;
    fixture.detectChanges();

    expect(fixture.debugElement
      .query(By.css('textarea'))
      .nativeElement
      .classList
    ).toContain('md-input--multiline');
  });


  it('should render filled class', () => {
    expect(groupEl.classList.contains('md-input--filled')).toBeFalsy();

    fixture.componentInstance.isFilled = true;
    fixture.detectChanges();

    expect(groupEl.classList.contains('md-input--filled')).toBe(true);
  });

  it('should render containerSize class', () => {
    expect(groupEl.classList.contains('small-6')).toBeFalsy();

    fixture.componentInstance.containerSize = 'small-6';
    fixture.detectChanges();

    expect(groupEl.classList.contains('small-6')).toBe(true);
  });

  it('should render inputSize class', () => {
    expect(wrapperEl.classList.contains('small-6')).toBeFalsy();

    fixture.componentInstance.inputSize = 'small-6';
    fixture.detectChanges();

    expect(wrapperEl.classList.contains('small-6')).toBe(true);
  });

  it('should render a disabled input', () => {
    expect(inputEl.disabled).toBe(false);

    fixture.componentInstance.disabled = true;
    fixture.detectChanges();

    expect(groupEl.classList.contains('md-disabled')).toBe(true);
    expect(inputEl.classList.contains('md-disabled')).toBe(true);
    expect(inputEl.disabled).toBe(true);
  });

  it('should render a placeholder', () => {
    expect(inputEl.placeholder).toBeFalsy();

    fixture.componentInstance.placeholder = 'test-placeholder';
    fixture.detectChanges();

    expect(inputEl.placeholder).toBe('test-placeholder');
  });

  it('should render a readonly input', () => {
    expect(inputEl.readOnly).toBeFalsy();

    fixture.componentInstance.readonly = true;
    fixture.detectChanges();

    expect(groupEl.classList.contains('md-read-only')).toBe(true);
    expect(inputEl.classList.contains('md-read-only')).toBe(true);
    expect(inputEl.readOnly).toBe(true);
  });

  it('should render a required input', () => {
    expect(inputEl.required).toBeFalsy();

    fixture.componentInstance.required = true;
    fixture.detectChanges();

    expect(inputEl.required).toBe(true);
  });

  it('should render a shape class', () => {
    expect(inputEl.classList.contains('md-input--pill')).toBeFalsy();

    fixture.componentInstance.shape = 'pill';
    fixture.detectChanges();

    expect(inputEl.classList.contains('md-input--pill')).toBe(true);
  });

  it('should render a specific type', () => {
    fixture.componentInstance.type = 'text';
    fixture.detectChanges();

    expect(inputEl.type).toBe('text');
  });

  it('should render a value', () => {
    expect(inputEl.value).toBe('undefined');

    fixture.componentInstance.value = 'test-value';
    fixture.detectChanges();

    expect(inputEl.value).toBe('test-value');
  });

  it('should handle focus', () => {
    expect(inputEl.classList.contains('md-active')).toBeFalsy();

    inputEl.focus();
    fixture.detectChanges();

    expect(inputEl.classList.contains('md-active')).toBe(true);
  });

  it('should handle clear prop', () => {
    expect(fixture.debugElement.query(By.css('md-icon'))).toBeFalsy();

    fixture.componentInstance.clear = true;
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-icon'))).toBeTruthy();
  });

  it('should handle clearAriaLabel prop', () => {
    fixture.componentInstance.clear = true;
    fixture.componentInstance.clearAriaLabel = 'test-clear';

    fixture.detectChanges();

    expect(
      fixture.debugElement
        .query(By.css('button'))
        .nativeElement
        .getAttribute('aria-label')
      ).toBe('test-clear');
  });

  it('should handle helpText prop', () => {
    fixture.componentInstance.helpText = 'test-help';

    fixture.detectChanges();

    expect(
      fixture.debugElement
        .query(By.css('md-input-helper'))
        .nativeElement
        .textContent
      ).toContain('test-help');
  });

  it('should handle label prop', () => {
    fixture.componentInstance.label = 'test-label';

    fixture.detectChanges();

    expect(
      fixture.debugElement
        .query(By.css('md-label'))
        .nativeElement
        .textContent
      ).toContain('test-label');
  });

  it('should handle secondary label prop', () => {
    fixture.componentInstance.secondaryLabel = 'test-sec-label';

    fixture.detectChanges();

    expect(
      fixture.debugElement
        .query(By.css('.md-input__secondary-label'))
        .nativeElement
        .textContent
      ).toContain('test-sec-label');
  });

  it('should handle nested level prop', () => {
    expect(groupEl.classList.contains('md-input--nested-2')).toBeFalsy();

    fixture.componentInstance.nestedLevel = 2;
    fixture.detectChanges();

    expect(groupEl.classList.contains('md-input--nested-2')).toBe(true);
  });

  it('should render inputBefore', () => {
    fixture.componentInstance.inputBefore = true;
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-section'))).toBeTruthy();
    expect(inputEl.classList.contains('md-input--before')).toBe(true);
  });

  it('should render inputAfter', () => {
    fixture.componentInstance.inputAfter = true;
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-section'))).toBeTruthy();
    expect(inputEl.classList.contains('md-input--after')).toBe(true);
  });

  it('should handle success messageArr', () => {
    fixture.componentInstance.messageArr = [{message: 'test', type: 'success'}];
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-message'))).toBeTruthy();
    expect(groupEl.classList.contains('md-success')).toBe(true);
  });

  it('should handle warning messageArr', () => {
    fixture.componentInstance.messageArr = [{message: 'test', type: 'warning'}];
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-message'))).toBeTruthy();
    expect(groupEl.classList.contains('md-warning')).toBe(true);
  });

  it('should handle error messageArr', () => {
    fixture.componentInstance.messageArr = [{message: 'test', type: 'error'}];
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-message'))).toBeTruthy();
    expect(groupEl.classList.contains('md-error')).toBe(true);
  });

  it('should handle success status', () => {
    fixture.componentInstance.status = 'success';
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-message'))).toBeTruthy();
    expect(groupEl.classList.contains('md-success')).toBe(true);
  });

  it('should handle warning status', () => {
    fixture.componentInstance.status = 'warning';
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-message'))).toBeTruthy();
    expect(groupEl.classList.contains('md-warning')).toBe(true);
  });

  it('should handle error status', () => {
    fixture.componentInstance.status = 'error';
    fixture.detectChanges();

    expect(fixture.debugElement.query(By.css('md-input-message'))).toBeTruthy();
    expect(groupEl.classList.contains('md-error')).toBe(true);
  });
});

