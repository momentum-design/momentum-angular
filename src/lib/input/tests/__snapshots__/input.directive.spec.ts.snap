// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputDirective should match snapshot 1`] = `
<ng-component>
  <md-input-group
    class="md-input-container"
  >
    
    <div
      class="md-input__wrapper"
      ng-reflect-klass="md-input__wrapper"
      ng-reflect-ng-class=""
    >
      
      
      <input
        class="md-input"
        id="test-id"
        mdinput=""
        ng-reflect-id="test-id"
      />
      
      
    </div>
    
    
    
  </md-input-group>
</ng-component>
`;
